File: Search/binary_search_recursive.py
def binary_search_recursive(array,target,start,end):
    if start > end:
        return None
    mid = (start + end) // 2
    #if mid is target
    if array[mid] == target:
        return mid
    #if target is smaller than mid
    elif array[mid] > target:
        return binary_search_recursive(array,target,start,mid-1)
    #if target is larger than mid
    else:
        return binary_search_recursive(array,target,mid+1,end)
    
arr = [1,2,3,4,5,6,7,8,9,10]

print(binary_search_recursive(arr,7,0,len(arr)-1))

File: Search/binary_search_recursive.java
import java.util.*;

//In order to use binary search, we should assume the array is sorted.
// binary search (recursive)
// runtime: O(log n)
public class binary_search_recursive {
  public static int binary_search_recursive(int[] arr, int target, int start, int end) {
    if (start > end) {
      return -1;
    }
    int mid = (start + end) / 2;
    if (arr[mid] == target) {
      return mid;
    }
    //if the target is less than the value of mid
    else if (arr[mid] > target) {
      return binary_search_recursive(arr,target,start, mid - 1);
    }
    //if the target is greater than the value of mid
    else {
      return binary_search_recursive(arr, target, mid + 1, end);
    }
  }
}
